pkg/graveler/graveler.go

```diff
@@ -1841,6 +1873,46 @@ func (g *Graveler) safeBranchWrite(ctx context.Context, log logging.Logger, repo
 	return nil
 }
 
+func (g *Graveler) UpdateObjectMetadata(ctx context.Context, repository *RepositoryRecord, branchID BranchID, key Key, update ValueUpdateFunc, opts ...SetOptionsFunc) error {
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repository, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+
+	options := NewSetOptions(opts)
+	if repository.ReadOnly && !options.Force {
+		return ErrReadOnlyRepository
+	}
+
+	log := g.log(ctx).WithFields(logging.Fields{"key": key, "operation": "update_user_metadata"})
+
+	err = g.safeBranchWrite(ctx, log, repository, branchID, safeBranchWriteOptions{MaxTries: options.MaxTries}, func(branch *Branch) error {
+		value, onStaging, err := g.getFromBranchWithStaging(ctx, repository, branch, key)
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return fmt.Errorf("get existing entry to update user metadata: %w", err)
+		}
+
+		return g.StagingManager.Update(ctx, branch.StagingToken, key, func(currentValue *Value) (*Value, error) {
+			if currentValue == nil {
+				if onStaging {
+					// Object previously on staging, and
+					// was deleted since
+					// getFromBranchWithStaging.
+					return nil, ErrNotFound
+				}
+				// Use committed value (in practice it came
+				// from committed, so it is non-nil).
+				currentValue = value
```

Comment:
If I understand correctly, the only reason we get the value before the update is because we perform retries. In that case I think it's important to document that so it will be clear what we're doing here.